confgenerator/feature.go

```diff
@@ -0,0 +1,246 @@
+package confgenerator
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// Errors returned by ExtractFeatures can be tested against these errors using
+// errors.Is
+var (
+	ErrTrackingInlineStruct   = errors.New("cannot have tracking on inline struct")
+	ErrTrackingOverrideStruct = errors.New("struct that has tracking tag must not be empty")
+	ErrMapAsField             = errors.New("map type for a field is not supported")
+	ErrTrackingOverrideMap    = errors.New("map that has tracking tag must not be empty")
+	ErrInvalidType            = errors.New("object in path must be of type Component")
+)
+
+type Feature struct {
+	// Module defines the sub-agent: metrics or logging
+	Module string
+	// Kind defines the kind: receivers or processors
+	Kind string
+	// Type from Component.Type()
+	Type string
+	// Key: set of keys that will be joined together for feature tracking metrics
+	Key []string
+	// Value defined from fields of UnifiedConfig.
+	Value string
+}
+
+// ExtractFeatures fields that containing a tracking tag will be tracked.
+// Automatic collection of bool or int fields. Any value that exists on tracking
+// tag will be used instead of value from UnifiedConfig.
+func ExtractFeatures(uc *UnifiedConfig) ([]Feature, error) {
+	allFeatures := getOverriddenDefaultPipelines(uc)
+	f, err := trackingFeatures(reflect.ValueOf(uc), metadata{}, Feature{})
```

Comment:
Can we unwrap and do some of the tracking of atleast the first 2 layers of `UnifiedConfig` here? instead of in `trackingFeatures`?

This is because `trackingFeatures` feels error prone to me as it relies on the depth of the recursion to build out `Feature.Kind` and `Feature.Kind`. Instead if we build those out here `trackingFeatures` could be a pure function with no assumptions being made about the depth of recursion.

Feel free to ignore this suggestion. I don't know yet which is more readable. My suggestion would add some duplication for sure in favor of readability but feel free to make the call yourself.